src/components/tl/mlx5/tl_mlx5_wqe.c

```diff
@@ -0,0 +1,304 @@
+                     sizeof(struct mlx5_wqe_mkey_context_seg) +
+                     sizeof(struct mlx5_wqe_umr_pointer_seg)) / DS_SIZE;
+    uint8_t fm_ce_se =
+        MLX5_WQE_CTRL_INITIATOR_SMALL_FENCE | MLX5_WQE_CTRL_CQ_UPDATE;
+    struct ibv_qp_ex *   qp_ex = ibv_qp_to_qp_ex(qp);
+    struct mlx5dv_qp_ex *mqp = mlx5dv_qp_ex_from_ibv_qp_ex(qp_ex);
+    struct mlx5_wqe_ctrl_seg *        ctrl;
+    struct mlx5_wqe_umr_ctrl_seg *    umr_ctrl_seg;
+    struct mlx5_wqe_mkey_context_seg *mk_seg;
+    struct mlx5_wqe_umr_pointer_seg * pseg;
+    char                              wqe_desc[n_ds * DS_SIZE];
+    int                               xlat_size;
+
+    ibv_wr_start(qp_ex);
+    memset(wqe_desc, 0, n_ds * DS_SIZE);
+
+    ctrl = (void *)wqe_desc;
+    mlx5dv_set_ctrl_seg(ctrl, /* pi */ 0x0, opcode, opmode, qp->qp_num,
+                        fm_ce_se, n_ds, 0x0, htobe32(dv_mkey->lkey));
+
+    umr_ctrl_seg = PTR_OFFSET(ctrl, sizeof(*ctrl));
+    umr_ctrl_seg->mkey_mask =
+        htobe64(MLX5_WQE_UMR_CTRL_MKEY_MASK_LEN |
+                MLX5_WQE_UMR_CTRL_MKEY_MASK_ACCESS_LOCAL_WRITE |
+                MLX5_WQE_UMR_CTRL_MKEY_MASK_ACCESS_REMOTE_READ |
+                MLX5_WQE_UMR_CTRL_MKEY_MASK_ACCESS_REMOTE_WRITE |
+                MLX5_WQE_UMR_CTRL_MKEY_MASK_ACCESS_ATOMIC |
+                MLX5_WQE_UMR_CTRL_MKEY_MASK_FREE);
+
+    mk_seg               = PTR_OFFSET(umr_ctrl_seg, sizeof(*umr_ctrl_seg));
+    mk_seg->access_flags = get_umr_mr_flags(access_flags);
+    mk_seg->qpn_mkey     = htobe32(0xffffff00 | (dv_mkey->lkey & 0xff));
+
+    pseg = PTR_OFFSET(mk_seg, sizeof(*mk_seg));
+
+    umr_pointer_seg_init(repeat_count, num_entries, data, pseg, ptr_mkey,
+                         ptr_address, &xlat_size, &reglen);
+
+    mk_seg->len = htobe64(reglen);
+    umr_ctrl_seg->klm_octowords =
+        htobe16(ucc_align_up(xlat_size, 64) / DS_SIZE);
+    mlx5dv_wr_raw_wqe(mqp, wqe_desc);
+    ibv_wr_complete(qp_ex);
+    return UCC_OK;
+}
+
+ucc_status_t ucc_tl_mlx5_post_rdma(struct ibv_qp *qp, uint32_t qpn,
+                                   struct ibv_ah *ah, uintptr_t src_mkey_addr,
+                                   size_t len, uint32_t src_mr_lkey,
+                                   uintptr_t dst_addr, uint32_t dst_mr_key,
+                                   int send_flags, uint64_t wr_id)
+
+{
+
+    uint32_t                      opcode = MLX5_OPCODE_RDMA_WRITE;
+    uint32_t                      opmode = 0x0;
+    struct ibv_qp_ex *            qp_ex = ibv_qp_to_qp_ex(qp);
+    struct mlx5dv_qp_ex *         mqp   = mlx5dv_qp_ex_from_ibv_qp_ex(qp_ex);
+    uint8_t                       fm_ce_se = MLX5_WQE_CTRL_INITIATOR_SMALL_FENCE;
+    struct mlx5_wqe_ctrl_seg *    ctrl;
+    struct mlx5_wqe_data_seg *    data;
+    struct mlx5_wqe_datagram_seg *dseg;
+    struct mlx5_wqe_raddr_seg *   rseg;
+    uint32_t n_ds = (sizeof(*ctrl) + (ah ? sizeof(*dseg) : 0) + sizeof(*rseg) +
```

Comment:
initialized vars first

Reply:
just a code style we stick to, if you need some uninit variables then please move n_ds initialization out of declaration block